Index: lcl/interfaces/gtk2/gtk2wsmenus.pp
===================================================================
--- lcl/interfaces/gtk2/gtk2wsmenus.pp	(revision 65368)
+++ lcl/interfaces/gtk2/gtk2wsmenus.pp	(working copy)
@@ -22,7 +22,7 @@
 
 uses
   // RTL
-  Classes, Types, glib2, gdk2, gtk2,
+  Classes, Types, glib2, gdk2, gtk2, math,
   // LazUtils
   LazTracer,
   // LCL
@@ -572,7 +572,8 @@
 var
   Requisition: TGtkRequisition;
   Alignment: TPopupAlignment;
-  ScreenHeight: gint;
+  MonitorNum: gint;
+  MonitorRect: TGdkRectangle;
 begin
   X^ := PPoint(WidgetInfo^.UserData)^.X;
   Y^ := PPoint(WidgetInfo^.UserData)^.Y;
@@ -579,15 +580,13 @@
 
   if WidgetInfo^.LCLObject is TPopupMenu then
   begin
-    // make menu to fit the screen vertically
     gtk_widget_size_request(PGtkWidget(menu), @Requisition);
-    ScreenHeight := gdk_screen_height();
-    if Y^ + Requisition.height > ScreenHeight then
-    begin
-      Y^ := ScreenHeight - Requisition.height;
-      if Y^ < 0 then Y^ := 0;
-    end;
 
+    // make menu to fit the monitor vertically
+    MonitorNum := gdk_screen_get_monitor_at_point(gdk_screen_get_default, X^, Y^);
+    gdk_screen_get_monitor_geometry(gdk_screen_get_default, MonitorNum, @MonitorRect);
+    Y^ := Max(MonitorRect.y, Min(Y^, MonitorRect.y + MonitorRect.height - Requisition.height));
+
     // get actual alignment
     Alignment := TPopupMenu(WidgetInfo^.LCLObject).Alignment;
     if TPopupMenu(WidgetInfo^.LCLObject).UseRightToLeftAlignment then
